src/main/java/org/springframework/integration/kafka/outbound/KafkaProducerMessageHandler.java

```diff
@@ -47,29 +57,65 @@
  *
  * @since 0.5
  */
-public class KafkaProducerMessageHandler<K, V> extends AbstractMessageHandler {
+public class KafkaProducerMessageHandler<K, V> extends AbstractReplyProducingMessageHandler {
 
-	private static final long DEFAULT_SEND_TIMEOUT = 10000;
+	private static final long DEFAULT_TIMEOUT = 10000;
 
 	private final KafkaTemplate<K, V> kafkaTemplate;
 
+	private final AbstractMessageListenerContainer<K, V> replyContainer;
+
+	private final Map<String, SettableListenableFuture<ConsumerRecord<K, V>>> pendingReplies =
```

Comment:
Since you made it String-based, we have to call `toString()` for the key everywhere:
![suspicious_call](https://cloud.githubusercontent.com/assets/1071590/24922064/f570cdaa-1eba-11e7-803b-cd91e0368f47.png)